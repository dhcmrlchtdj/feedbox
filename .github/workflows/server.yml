name: feedbox-server

on:
    push:
        branches:
            - master
        paths:
            - '.github/workflows/server.yml'
            - 'server/**'

jobs:
    build-deploy:
        runs-on: ubuntu-18.04
        steps:
            - uses: actions/checkout@v2.0.0
            - name: push to heroku
              run: git subtree push --prefix=server https://heroku:$HEROKU_AUTH_TOKEN@git.heroku.com/$HEROKU_APP_NAME.git master
              env:
                  HEROKU_APP_NAME: "fbox"
                  HEROKU_AUTH_TOKEN: ${{ secrets.HEROKU_AUTH_TOKEN }}
