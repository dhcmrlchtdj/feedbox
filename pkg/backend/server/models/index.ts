import { db } from './connection'

export const model = {
    async destroy() {
        db.$pool.end()
    },

    async getUserById(id: number): Promise<User | null> {
        const user = await db.oneOrNone<User>(
            `SELECT id, platform, pid, addition
            FROM users
            WHERE id = $1`,
            id,
        )
        return user
    },

    async getOrCreateUserByGithub(
        githubId: string,
        email: string,
    ): Promise<User> {
        return db.tx(async (t) => {
            let user = await t.oneOrNone<GithubUser>(
                `SELECT id, platform, pid, addition
                FROM users
                WHERE platform = 'github'
                AND pid = $1`,
                [githubId],
            )
            if (user === null) {
                user = await t.one<GithubUser>(
                    `INSERT INTO users(platform, pid, addition)
                    VALUES ('github', $1, $2)
                    RETURNING id, platform, pid, addition`,
                    [githubId, { email }],
                )
            } else {
                if (user.addition.email !== email) {
                    await t.none(
                        `UPDATE users SET addition = $1 WHERE id = $2`,
                        [{ email }, user.id],
                    )
                }
            }
            return user
        })
    },

    async getOrCreateUserByTelegram(chatId: string): Promise<User> {
        return db.tx(async (t) => {
            let user = await t.oneOrNone<TelegramUser>(
                `SELECT id, platform, pid, addition
                FROM users
                WHERE platform = 'telegram'
                AND pid = $1`,
                [chatId],
            )
            if (user === null) {
                user = await t.one<TelegramUser>(
                    `INSERT INTO users(platform, pid)
                    VALUES ('telegram', $1)
                    RETURNING id, platform, pid, addition`,
                    [chatId],
                )
            }
            return user
        })
    },

    async getFeedIdByUrl(url: string): Promise<number> {
        return db.tx(async (t) => {
            let feed = await t.oneOrNone<{ id: number }>(
                `SELECT id
                FROM feeds
                WHERE url = $1`,
                [url],
            )
            if (feed === null) {
                feed = await t.one<{ id: number }>(
                    `INSERT INTO feeds(url) VALUES ($1) RETURNING id`,
                    [url],
                )
            }
            return feed.id
        })
    },

    async getFeedByUser(userId: number): Promise<Feed[]> {
        return db.manyOrNone<Feed>(
            `SELECT id, url, updated
            FROM feeds
            WHERE id IN (SELECT fid FROM r_user_feed WHERE uid = $1)
            ORDER BY updated DESC`,
            [userId],
        )
    },

    async updateFeed(
        id: number,
        links: string[],
        updated: Date,
    ): Promise<void> {
        await db.none(
            `UPDATE feeds
            SET links = $1, updated = $2
            WHERE id = $3`,
            [JSON.stringify(links), updated, id],
        )
    },

    async getActiveFeeds(): Promise<Feed[]> {
        return db.manyOrNone<Feed>(
            `SELECT id, url, updated
            FROM feeds
            WHERE id IN (SELECT fid FROM r_user_feed)`,
        )
    },

    async getLinks(feedId: number): Promise<string[]> {
        const feed = await db.one<{ links: string[] }>(
            `SELECT links
            FROM feeds
            WHERE id = $1`,
            [feedId],
        )
        return feed.links
    },

    async getSubscribers(feedId: number): Promise<User[]> {
        return db.manyOrNone<User>(
            `SELECT id, platform, pid, addition
            FROM users
            WHERE id IN (SELECT uid FROM r_user_feed WHERE fid = $1)`,
            feedId,
        )
    },

    async subscribe(userId: number, feedId: number): Promise<number> {
        const r = await db.result(
            `INSERT INTO r_user_feed(uid, fid)
            VALUES ($1, $2)
            ON CONFLICT DO NOTHING`,
            [userId, feedId],
        )
        return r.rowCount
    },

    async unsubscribe(userId: number, feedId: number): Promise<number> {
        const r = await db.result(
            `DELETE FROM r_user_feed
            WHERE uid = $1
            AND fid = $2`,
            [userId, feedId],
        )
        return r.rowCount
    },

    async subscribeUrls(userId: number, urls: string[]): Promise<number> {
        if (urls.length === 0) return 0
        return db.tx(async (t) => {
            const fvalues = urls.map((_, idx) => `(\$${idx + 1})`).join(', ')
            await t.none(
                `INSERT INTO feeds(url) VALUES ${fvalues}
                ON CONFLICT DO NOTHING`,
                urls,
            )

            const rvalues = urls.map((_, idx) => `\$${idx + 2}`).join(', ')
            const r = await t.result(
                `INSERT INTO r_user_feed(uid, fid)
                SELECT $1, id FROM feeds WHERE url IN (${rvalues})
                ON CONFLICT DO NOTHING`,
                [userId, ...urls],
            )
            return r.rowCount
        })
    },

    async unsubscribeUrls(userId: number, urls: string[]): Promise<number> {
        if (urls.length === 0) return 0
        const uvalues = urls.map((_, idx) => `\$${idx + 2}`).join(', ')
        const r = await db.result(
            `DELETE FROM r_user_feed
            WHERE uid = $1
            AND fid IN (SELECT id FROM feeds WHERE url in (${uvalues}))`,
            [userId, ...urls],
        )
        return r.rowCount
    },

    async unsubscribeAll(userId: number): Promise<number> {
        const r = await db.result(
            `DELETE FROM r_user_feed
            WHERE uid = $1`,
            [userId],
        )
        return r.rowCount
    },
}

export type GithubUser = {
    id: number
    platform: 'github'
    pid: string
    addition: {
        email: string
    }
}
export type TelegramUser = {
    id: number
    platform: 'telegram'
    pid: string
    addition: {}
}
export type User = GithubUser | TelegramUser

export type Feed = {
    id: number
    url: string
    updated: number | null
}
