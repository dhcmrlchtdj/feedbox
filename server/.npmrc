disturl=https://npm.taobao.org/mirrors/node
fse_binary_host_mirror=https://npm.taobao.org/mirrors/fsevents
node_sqlite3_binary_host_mirror=https://npm.taobao.org/mirrors
