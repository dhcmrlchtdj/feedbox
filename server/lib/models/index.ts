import * as Knex from 'knex'
import cfg from './cfg'
import lazy from '../utils/lazy'

const conn = lazy(() => Knex(cfg))

export default {
    async init() {},
    async destroy() {
        await conn().destroy()
    },

    async getUserById(id: number): Promise<User | null> {
        const r = await conn()
            .select('id', 'github_id as githubId', 'email')
            .from('feedbox_user')
            .where({ id })
        if (r.length > 0) {
            return r[0]
        } else {
            return null
        }
    },

    async getUserIdByGithub(
        github_id: number,
        email: string,
    ): Promise<number | null> {
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_user(github_id,email) VALUES(?,?)
                ON CONFLICT(github_id) DO UPDATE SET email=?`,
                [github_id, email, email],
            )
            const r = await tnx
                .select('id')
                .from('feedbox_user')
                .where({ email })
            await tnx.commit()
            return r[0].id
        } catch (err) {
            await tnx.rollback(err)
            return null
        }
    },

    async getFeedIdByUrl(url: string): Promise<number | null> {
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_feed(url) VALUES(?)
                ON CONFLICT DO NOTHING`,
                [url],
            )
            const r = await tnx
                .select('id')
                .from('feedbox_feed')
                .where({ url })
            await tnx.commit()
            return r[0].id
        } catch (err) {
            await tnx.rollback(err)
            return null
        }
    },

    async getFeedByUser(userId: number): Promise<Feed[]> {
        const r = await conn()
            .select(
                'feedbox_feed.id as id',
                'feedbox_feed.url as url',
                'feedbox_feed.updated as updated',
            )
            .from('feedbox_r_user_feed')
            .innerJoin(
                'feedbox_user',
                'feedbox_user.id',
                'feedbox_r_user_feed.user_id',
            )
            .innerJoin(
                'feedbox_feed',
                'feedbox_feed.id',
                'feedbox_r_user_feed.feed_id',
            )
            .where('feedbox_user.id', userId)
            .orderBy('feedbox_feed.updated', 'desc')
        return r
    },

    async prepareFeedForUpdate(): Promise<FeedDoc[]> {
        const feeds = await conn()
            .select(
                'feedbox_feed.id as id',
                'feedbox_feed.url as url',
                'feedbox_feed.updated as updated',
                'feedbox_user.email as email',
            )
            .from('feedbox_r_user_feed')
            .innerJoin(
                'feedbox_feed',
                'feedbox_feed.id',
                'feedbox_r_user_feed.feed_id',
            )
            .innerJoin(
                'feedbox_user',
                'feedbox_user.id',
                'feedbox_r_user_feed.user_id',
            )
        const map = new Map<number, FeedDoc>()
        feeds.forEach(({ id, url, updated, email }) => {
            const v = map.get(id) ?? {
                id,
                url,
                updated,
                emails: [] as string[],
                links: new Set(),
            }
            v.emails.push(email)
            map.set(id, v)
        })
        const links = await conn()
            .select('feed_id', 'url')
            .from('feedbox_link')
            .whereIn('feed_id', Array.from(map.keys()))
        links.forEach(({ feed_id, url }) => {
            const v = map.get(feed_id)!
            v.links.add(url)
        })
        return Array.from(map.values())
    },

    async addLinks(links: { feed_id: number; url: string }[]) {
        if (links.length === 0) return
        await conn()
            .insert(links)
            .into('feedbox_link')
    },

    async updateFeedUpdated(id: number, updated: Date) {
        await conn()('feedbox_feed')
            .where({ id })
            .update({ updated })
    },

    async subscribe(user_id: number, feed_id: number) {
        await conn().raw(
            `INSERT INTO feedbox_r_user_feed(user_id, feed_id) VALUES(?,?)
            ON CONFLICT DO NOTHING`,
            [user_id, feed_id],
        )
    },

    async unsubscribe(user_id: number, feed_id: number) {
        await conn()
            .from('feedbox_r_user_feed')
            .where({ user_id, feed_id })
            .del()
    },

    async subscribeUrls(user_id: number, urls: string[]) {
        if (urls.length === 0) return
        const qvalue = urls.map(_ => '(?)').join(',')
        const qrange = urls.map(_ => '?').join(',')
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_feed(url) VALUES ${qvalue}
                ON CONFLICT DO NOTHING`,
                urls,
            )
            await tnx.raw(
                `INSERT INTO feedbox_r_user_feed(user_id, feed_id)
                SELECT ?, feedbox_feed.id FROM feedbox_feed WHERE feedbox_feed.url in (${qrange})
                ON CONFLICT DO NOTHING`,
                [user_id, ...urls],
            )
            await tnx.commit()
        } catch (err) {
            await tnx.rollback(err)
        }
    },
}

export interface User {
    id: number
    github_id: number
    email: string
}

export interface Feed {
    id: number
    url: string
    updated: number | null
}

export interface Link {
    id: number
    url: string
    feed_id: number
}

export interface RUserFeed {
    user_id: number
    feed_id: number
}

export interface FeedDoc {
    id: number
    url: string
    updated: number | null
    emails: string[]
    links: Set<string>
}
