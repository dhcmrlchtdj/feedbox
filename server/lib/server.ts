import * as Hapi from '@hapi/hapi'
import plugins from './plugins'
import routes from './routes'
import prepare from './prepare'
import Model from './models'

const common = async () => {
    await prepare()

    const corsConf = {
        headers: [
            'accept',
            'authorization',
            'content-type',
            'if-none-match',
            'content-type',
            'x-sw-strategy',
            'x-sw-race',
            'x-sw-actions',
        ],
        credentials: true,
    }
    const server = Hapi.server({
        port: Number(process.env.PORT || 8000),
        host: '0.0.0.0',
        routes: {
            payload: { allow: 'application/json' },
            cors: process.env.SERVER === process.env.WEB ? false : corsConf,
        },
    })
    await plugins(server)
    server.route(routes)

    server.events.on('stop', async () => {
        await Model.destroy()
    })

    return server
}

export const init = async () => {
    const server = await common()
    server.initialize()
    return server
}

export const start = async () => {
    const server = await common()
    server.start()
    return server
}
