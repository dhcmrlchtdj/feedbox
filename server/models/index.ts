import Knex from 'knex'
import { config } from './config'
import { lazy } from '../../util/lazy'

const conn = lazy(() => Knex(config))

export const model = {
    async init() {},
    async destroy() {
        await conn().destroy()
    },

    async getUserById(id: number): Promise<User | null> {
        const r = await conn()
            .select('id', 'github_id as githubId', 'email')
            .from('feedbox_user')
            .where({ id })
        if (r.length > 0) {
            return r[0]
        } else {
            return null
        }
    },

    async getUserIdByGithub(
        github_id: number,
        email: string,
    ): Promise<number | null> {
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_user(github_id,email) VALUES(?,?)
                ON CONFLICT(github_id) DO UPDATE SET email=?`,
                [github_id, email, email],
            )
            const r = await tnx
                .select('id')
                .from('feedbox_user')
                .where({ email })
            await tnx.commit()
            return r[0].id
        } catch (err) {
            await tnx.rollback(err)
            return null
        }
    },

    async getFeedIdByUrl(url: string): Promise<number | null> {
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_feed(url) VALUES(?)
                ON CONFLICT DO NOTHING`,
                [url],
            )
            const r = await tnx.select('id').from('feedbox_feed').where({ url })
            await tnx.commit()
            return r[0].id
        } catch (err) {
            await tnx.rollback(err)
            return null
        }
    },

    async getFeedByUser(userId: number): Promise<Feed[]> {
        const r = await conn()
            .select(
                'feedbox_feed.id as id',
                'feedbox_feed.url as url',
                'feedbox_feed.updated as updated',
            )
            .from('feedbox_r_user_feed')
            .innerJoin(
                'feedbox_user',
                'feedbox_user.id',
                'feedbox_r_user_feed.user_id',
            )
            .innerJoin(
                'feedbox_feed',
                'feedbox_feed.id',
                'feedbox_r_user_feed.feed_id',
            )
            .where('feedbox_user.id', userId)
            .orderBy('feedbox_feed.updated', 'desc')
        return r
    },

    async getActiveFeeds(): Promise<Feed[]> {
        const feeds = await conn()
            .select(
                'feedbox_feed.id as id',
                'feedbox_feed.url as url',
                'feedbox_feed.links as links',
                'feedbox_feed.updated as updated',
            )
            .from('feedbox_feed')
            .innerJoin(
                'feedbox_r_user_feed',
                'feedbox_r_user_feed.feed_id',
                'feedbox_feed.id',
            )
        return feeds
    },

    async getSubscribers(id: number): Promise<User[]> {
        const users = await conn()
            .select(
                'feedbox_user.id as id',
                'feedbox_user.github_id as githubId',
                'feedbox_user.email as email',
            )
            .from('feedbox_user')
            .innerJoin(
                'feedbox_r_user_feed',
                'feedbox_r_user_feed.user_id',
                'feedbox_user.id',
            )
            .where({ 'feedbox_r_user_feed.feed_id': id })
        return users
    },

    async updateFeed(id: number, links: string[], updated: Date) {
        await conn()('feedbox_feed')
            .where({ id })
            .update({ links: JSON.stringify(links), updated })
    },

    async subscribe(user_id: number, feed_id: number) {
        await conn().raw(
            `INSERT INTO feedbox_r_user_feed(user_id, feed_id) VALUES(?,?)
            ON CONFLICT DO NOTHING`,
            [user_id, feed_id],
        )
    },

    async unsubscribe(user_id: number, feed_id: number) {
        await conn()
            .from('feedbox_r_user_feed')
            .where({ user_id, feed_id })
            .del()
    },

    async subscribeUrls(user_id: number, urls: string[]) {
        if (urls.length === 0) return
        const qvalue = urls.map((_) => '(?)').join(',')
        const qrange = urls.map((_) => '?').join(',')
        const tnx = await conn().transaction()
        try {
            await tnx.raw(
                `INSERT INTO feedbox_feed(url) VALUES ${qvalue}
                ON CONFLICT DO NOTHING`,
                urls,
            )
            await tnx.raw(
                `INSERT INTO feedbox_r_user_feed(user_id, feed_id)
                SELECT ?, feedbox_feed.id FROM feedbox_feed WHERE feedbox_feed.url in (${qrange})
                ON CONFLICT DO NOTHING`,
                [user_id, ...urls],
            )
            await tnx.commit()
        } catch (err) {
            await tnx.rollback(err)
        }
    },
}

export interface User {
    id: number
    github_id: number
    email: string
}

export interface Feed {
    id: number
    url: string
    updated: number | null
    links: string[]
}

export interface RUserFeed {
    user_id: number
    feed_id: number
}
