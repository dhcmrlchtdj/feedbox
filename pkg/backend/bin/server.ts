import 'newrelic'
import { start } from '../server/server'

start()
