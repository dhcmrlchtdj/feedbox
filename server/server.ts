import './init-env'
import * as path from 'path'
import * as Hapi from '@hapi/hapi'
import { addPlugins } from './plugins'
import { routes } from './routes'
import { model } from './models'
import { telegramBot } from './telegram/bot'

const common = async () => {
    const server = Hapi.server({
        port: Number(process.env.PORT || 8000),
        host: '0.0.0.0',
        routes: {
            payload: { allow: 'application/json' },
            files: {
                relativeTo: path.resolve(__dirname, '../static'),
            },
        },
    })
    await addPlugins(server)
    server.route(routes)

    server.events.on('stop', async () => {
        await model.destroy()
    })

    return server
}

export const init = async () => {
    const server = await common()
    server.initialize()
    return server
}

export const start = async () => {
    const server = await common()
    server.start()

    if (process.env.NODE_ENV === 'production') {
        // init telegram webhook
        await telegramBot.registerWebhook()
    }

    return server
}
