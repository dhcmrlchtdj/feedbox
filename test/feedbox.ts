import FormData from 'form-data'
import { init } from '../server/server'
import { model } from '../server/models'

let server: any
const auth = {
    strategy: 'session',
    credentials: { userId: 0 },
}
beforeAll(async () => {
    server = await init()
    const id = await model.getUserIdByGithub(1, 'user@example.com')
    if (id === null) {
        throw Error('user id')
    } else {
        auth.credentials.userId = id
    }
})
afterAll(async () => {
    if (server) {
        await server.stop()
    }
})

describe('user API', () => {
    test('/api/v1/user', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/user',
            auth,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('/api/logout', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/logout',
            headers: {
                Cookie: 'token=token',
            },
        })
        expect(resp.headers['set-cookie']).toMatchSnapshot()
        expect(resp.headers['location']).toMatchSnapshot()
        expect(resp.statusCode).toMatchSnapshot()
    })
})

describe('feed API', () => {
    test('/api/v1/feeds/add', async () => {
        const resp = await server.inject({
            method: 'put',
            url: '/api/v1/feeds/add',
            payload: {
                url: 'https://example.com/rss',
            },
            auth,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('/api/v1/feeds', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/feeds',
            auth,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('/api/v1/feeds/export', async () => {
        const resp = await server.inject({
            method: 'get',
            url: '/api/v1/feeds/export',
            auth,
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('/api/v1/feeds/import', async () => {
        const form = new FormData()
        const opml = `
            <?xml version="1.0" encoding="utf-8"?>
            <opml version="1.0">
            <head><title>feeds</title></head>
            <body>
            <outline type="rss" text="example" xmlUrl="https://example.com/rss2"/>
            <outline type="rss" text="example" xmlUrl="https://example.com/rss3"/>
            </body>
            </opml>
        `
        form.append('opml', opml, {
            filename: 'feed.opml',
            contentType: 'text/x-opml',
        })
        const resp = await server.inject({
            method: 'post',
            url: '/api/v1/feeds/import',
            auth,
            payload: form.getBuffer(),
            headers: form.getHeaders(),
        })
        expect(resp.result).toMatchSnapshot()
    })

    test('/api/v1/feeds/remove', async () => {
        const resp = await server.inject({
            method: 'delete',
            url: '/api/v1/feeds/remove',
            payload: {
                feedId: 1,
            },
            auth,
        })
        expect(resp.result).toMatchSnapshot()
    })
})

describe('model', () => {
    test('prepareFeedForUpdate', async () => {
        const feed = await model.prepareFeedForUpdate()
        expect(feed).toMatchSnapshot()
    })
    test('updateFeedUpdated', async () => {
        await model.updateFeedUpdated(3, new Date(1580601600000))
        const curr = await model.getFeedByUser(1)
        expect(curr).toMatchSnapshot()
    })
})
