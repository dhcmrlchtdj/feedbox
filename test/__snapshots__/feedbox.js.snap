// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feed API /api/v1/feeds 1`] = `
Array [
  Object {
    "id": 1,
    "updated": null,
    "url": "https://example.com/rss",
  },
]
`;

exports[`feed API /api/v1/feeds/add 1`] = `
Array [
  Object {
    "id": 1,
    "updated": null,
    "url": "https://example.com/rss",
  },
]
`;

exports[`feed API /api/v1/feeds/export 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<opml version=\\"1.0\\">
<head><title>feeds</title></head>
<body>
<outline type=\\"rss\\" text=\\"example.com\\" xmlUrl=\\"https://example.com/rss\\"/>
</body>
</opml>"
`;

exports[`feed API /api/v1/feeds/import 1`] = `
Array [
  Object {
    "id": 1,
    "updated": null,
    "url": "https://example.com/rss",
  },
  Object {
    "id": 2,
    "updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "updated": null,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`feed API /api/v1/feeds/remove 1`] = `
Array [
  Object {
    "id": 2,
    "updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "updated": null,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`model getActiveFeeds 1`] = `
Array [
  Object {
    "id": 2,
    "links": Array [],
    "updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "links": Array [
      "feed1",
      "feed2",
    ],
    "updated": 2020-02-02T00:00:00.000Z,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`model updateFeed 1`] = `
Array [
  Object {
    "id": 2,
    "updated": null,
    "url": "https://example.com/rss2",
  },
  Object {
    "id": 3,
    "updated": 2020-02-02T00:00:00.000Z,
    "url": "https://example.com/rss3",
  },
]
`;

exports[`user API /api/logout 1`] = `
Array [
  "token=; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT; HttpOnly; SameSite=Strict; Path=/api",
]
`;

exports[`user API /api/logout 2`] = `"http://localhost:8000"`;

exports[`user API /api/logout 3`] = `302`;

exports[`user API /api/v1/user 1`] = `
Object {
  "email": "user@example.com",
  "githubId": 1,
  "id": 1,
}
`;
